Library/Homebrew/keg_relocate.rb

```diff
@@ -44,12 +57,106 @@ def replace_text!(text)
 
       any_changed = T.let(nil, T.nilable(String))
       sorted_keys.each do |key|
-        changed = text.gsub!(key, replacements.fetch(key))
+        replacement = replacements.fetch(key)
+        # Use selective replacement for /usr/local paths when HOMEBREW_PREFIX=/usr/local
+        changed = if should_use_selective_usr_local_replacement?(key, replacement)
+          selective_replace_usr_local!(T.cast(key, Regexp), replacement, text)
+        else
+          text.gsub!(key, replacement)
+        end
         any_changed ||= changed
       end
       !any_changed.nil?
     end
 
+    private
+
+    sig { params(pattern: T.any(String, Regexp), replacement: String).returns(T::Boolean) }
+    def should_use_selective_usr_local_replacement?(pattern, replacement)
+      # Only use selective replacement when:
+      # 1. HOMEBREW_PREFIX is /usr/local
+      # 2. We're replacing with @@HOMEBREW_PREFIX@@
+      # 3. The pattern matches /usr/local
+      # 4. Formula hasn't disabled this via JSON config
+      return false unless HOMEBREW_PREFIX.to_s == "/usr/local"
+      return false unless replacement == "@@HOMEBREW_PREFIX@@"
+      return false unless pattern.is_a?(Regexp)
+      return false unless pattern.source.include?("/usr/local")
+      return false if formula_disables_selective_replacement?
+
+      true
+    end
+
+    sig { returns(T::Boolean) }
+    def formula_disables_selective_replacement?
+      # Check if there's a JSON config file that disables selective replacement
+      # This allows per-formula overrides if issues arise
+      return false unless @keg_path
+
+      formula_name = @keg_path.basename.to_s
+      
+      # Look for a JSON config in the tap directory structure
+      # The config file should be at <tap>/selective_replacement_disabled.json
+      possible_tap_paths = [
+        @keg_path.parent.parent.parent / "selective_replacement_disabled.json", # For homebrew-core
+        @keg_path.parent.parent / "selective_replacement_disabled.json", # Alternative location
+      ]
+
+      possible_tap_paths.each do |config_path|
+        next unless config_path.exist?
+
+        begin
+          require "json"
+          config = JSON.parse(config_path.read)
+          disabled_formulas = config["disabled_formulas"] || []
+          return true if disabled_formulas.include?(formula_name)
+        rescue JSON::ParserError, Errno::ENOENT
+          # Ignore parsing errors and continue
+          next
+        end
+      end
+
+      false
+    end
+
+    sig { params(pattern: Regexp, replacement: String, text: String).returns(T.nilable(String)) }
+    def selective_replace_usr_local!(pattern, replacement, text)
```

Comment:
Lean into the existing logic more heavily, avoid replicating it whenever possible. This should be as minimal and DRY as possible. A better strategy would be to make multiple new replacement calls to the existing logic but instead of using `/usr/local` and `@@HOMEBREW_PREFIX@@` instead do e.g. `/usr/local/opt` and `@@HOMEBREW_PREFIX@@/opt`. That should avoid the need for any new `selective_replace_usr_local` function at all.